### PR TITLE
In agent.stream() and agent.astream() yield actions before observations

### PR DESCRIPTION
**Pull request by nfcampos**
- This emits actions as soon as they are available, ie. right after plan
- When action is executed an ActionObservation object is emitted, which contains both the action and the return value of the tool

```
---- chunk 1
{'actions': [AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})])]}
---- chunk 2
{'observations': [AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})]), observation="Vittoria Ceretti's")]}
---- chunk 3
{'actions': [AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})])]}
---- chunk 4
{'observations': [AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})]), observation='25 years')]}
---- chunk 5
{'actions': [AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})])]}
---- chunk 6
{'observations': [AgentObservation(action=AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})]), observation='Answer: 3.991298452658078')]}
---- chunk 7
{'output': "Leo DiCaprio's current girlfriend is Vittoria Ceretti, and her age "
           '(25 years) raised to the 0.43 power is approximately 3.99.'}
---- final
{'actions': [AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})]),
             AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})]),
             AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})])],
 'observations': [AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Leo DiCaprio's current girlfriend", log="\nInvoking: `Search` with `Leo DiCaprio's current girlfriend`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Leo DiCaprio\'s current girlfriend"\n}'}})]), observation="Vittoria Ceretti's"),
                  AgentObservation(action=AgentActionMessageLog(tool='Search', tool_input="Vittoria Ceretti's current age", log="\nInvoking: `Search` with `Vittoria Ceretti's current age`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Search', 'arguments': '{\n  "__arg1": "Vittoria Ceretti\'s current age"\n}'}})]), observation='25 years'),
                  AgentObservation(action=AgentActionMessageLog(tool='Calculator', tool_input='25^0.43', log='\nInvoking: `Calculator` with `25^0.43`\n\n\n', message_log=[AIMessageChunk(content='', additional_kwargs={'function_call': {'name': 'Calculator', 'arguments': '{\n  "__arg1": "25^0.43"\n}'}})]), observation='Answer: 3.991298452658078')],
 'output': "Leo DiCaprio's current girlfriend is Vittoria Ceretti, and her age "
           '(25 years) raised to the 0.43 power is approximately 3.99.'}
```

### CODE REVIEW

File: libs/langchain/langchain/agents/agent.py [hunk 1]
```diff
@@ -925,6 +942,27 @@ def _take_next_step(
         intermediate_steps: List[Tuple[AgentAction, str]],
         run_manager: Optional[CallbackManagerForChainRun] = None,
     ) -> Union[AgentFinish, List[Tuple[AgentAction, str]]]:
+        return self._consume_next_step(
+            [
+                a
```
Comment (eyurtsev) [inline libs/langchain/langchain/agents/agent.py:945]: a -> next_step?? a better variable name would be useful

File: libs/langchain/langchain/schema/agent.py [hunk 2]
```diff
@@ -50,6 +50,15 @@ class AgentActionMessageLog(AgentAction):
     type: Literal["AgentActionMessageLog"] = "AgentActionMessageLog"  # type: ignore
 
 
+class AgentObservation(Serializable):
```
Comment (hwchase17) [inline libs/langchain/langchain/schema/agent.py:53]: AgentObservation.observation seems weird namespacing

maybe AgentTurn? AgentIteration? AgentStep?

File: libs/langchain/langchain/agents/agent.py [hunk 3]
```diff
@@ -1026,6 +1066,27 @@ async def _atake_next_step(
         intermediate_steps: List[Tuple[AgentAction, str]],
         run_manager: Optional[AsyncCallbackManagerForChainRun] = None,
     ) -> Union[AgentFinish, List[Tuple[AgentAction, str]]]:
+        return self._consume_next_step(
+            [
+                a
+                async for a in self._aiter_next_step(
+                    name_to_tool_map,
```
Comment (eyurtsev) [inline libs/langchain/langchain/agents/agent.py:1073]: variable name

**Review by eyurtsev — APPROVED** [review]
Looks reasonable just minor nits for variable name

**Review by hwchase17 — APPROVED** [review]

**Pull request merged**